### PR TITLE
Add support for mremap MREMAP_DONTUNMAP flag

### PR DESCRIPTION
**Pull request by laipuran**
### Summary

- **Feature**: Implement the `MREMAP_DONTUNMAP` flag for the `mremap` syscall (Linux 5.7+). When combined with `MREMAP_MAYMOVE`, physical pages are moved to a new virtual address while the old address remains mapped as anonymous zero-fill-on-demand memory. A new `RemapOldMappingAction` enum (`Keep`/`Unmap`) replaces a bare boolean parameter for self-documenting call sites.

- **Refactor**: Replace the `dont_unmap: bool` parameter of `Vmar::remap()` with a `RemapOldMappingAction` enum. Move `is_userspace_vaddr_range` from `fn` to `pub(crate)` for use by the syscall handler.

- **Fix**: Add `new_addr` page-alignment and userspace-range validation at the syscall boundary when `MREMAP_FIXED` is set, catching invalid arguments early with `EINVAL` rather than descending into VM internals. Update the outdated FIXME comment that incorrectly claimed `resize_mapping` returns `EACCES`.

- **Test**: Add 6 regression-test cases for `MREMAP_DONTUNMAP` covering basic usage (content moved, old address zero-filled), error paths (without `MAYMOVE`, mismatched sizes), and combined usage with `MREMAP_FIXED`. Unblock 3 gVisor `Copy`-variant tests (`InPlace_Copy`, `MayMove_Copy`, `MustMove_Copy`) from the blocklist.

- **Docs**: Add the `MREMAP_DONTUNMAP` flag variant to the SCML compatibility documentation.

### CODE REVIEW

File: book/src/kernel/linux-compatibility/syscall-flag-coverage/memory-management/mremap.scml [hunk 1]
```diff
@@ -14,3 +14,12 @@ mremap(
...
 );
+
+// Relocate an existing memory mapping while keeping the old mapping
+// as anonymous zero-fill-on-demand.
+mremap(
+    old_address,
+    old_size,
+    new_size,
+    flags = MREMAP_MAYMOVE | MREMAP_DONTUNMAP
+);
```
Comment (cchanging) [inline book/src/kernel/linux-compatibility/syscall-flag-coverage/memory-management/mremap.scml]: This modification can be apply to line 6 directly.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 2]
```diff
@@ -212,6 +234,28 @@ impl Vmar {
...
+                    "remap: old_size must be non-zero when keeping the old mapping",
+                )
+            })?;
+            let anon_mapping = VmMapping::new(
+                old_size_nz,
+                old_addr,
+                MappedMemory::Anonymous,
+                None,
+                false,
+                true,
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: I don't think this is correct. In my understanding, the semantic is keeping the VMA for the old address range intact. So the old address range remains mapped, but the pages that were originally resident there have been moved away.

Or am I misunderstanding it? Can you provide some references?

**Review by cchanging — COMMENTED** [review]
Thanks for trying to implement the `MREMAP_DONTUNMAP` flag！ I took a quick look at the implementation, and it seems the semantics don't quite match my understanding. Could you add an extra test for `MREMAP_DONTUNMAP` with file-backed mappings?

File: kernel/src/vm/vmar/vmar_impls/remap.rs (hunk 2 above)
Comment (laipuran) [reply kernel/src/vm/vmar/vmar_impls/remap.rs]: It seems like i got a wrong understanding from the manual (

After reading https://man7.org/linux/man-pages/man2/mremap.2.html, i'm correcting and will push later.

File: kernel/src/syscall/mremap.rs [hunk 3]
```diff
@@ -51,10 +60,46 @@ fn do_sys_mremap(
     let old_size = old_size.align_up(PAGE_SIZE);
     let new_size = new_size.align_up(PAGE_SIZE);
 
+    // MREMAP_DONTUNMAP requires MREMAP_MAYMOVE and old_size == new_size.
```
Comment (cchanging) [inline kernel/src/syscall/mremap.rs]: MREMAP_DONTUNMAP -> `MREMAP_DONTUNMAP`; MREMAP_MAYMOVE -> `MREMAP_MAYMOVE`; old_size == new_size => `old_size == new_size`. Please check the other document modifications in this PR.

File: kernel/src/syscall/mremap.rs [hunk 4]
```diff
@@ -51,10 +60,46 @@ fn do_sys_mremap(
...
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "mremap: `MREMAP_DONTUNMAP` must be combined with `MREMAP_MAYMOVE`"
+            );
+        }
+        if new_size != old_size {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "mremap: `MREMAP_DONTUNMAP` requires `new_size` equals `old_size`"
+            );
```
Comment (cchanging) [inline kernel/src/syscall/mremap.rs:64]: We can move this check to the `if flags.contains(MremapFlags::MREMAP_DONTUNMAP)` block?

Besides, does this check verify the size after alignment?

File: kernel/src/syscall/mremap.rs [hunk 5]
```diff
@@ -51,10 +60,46 @@ fn do_sys_mremap(
...
+    // `new_addr`.  Reject early with EINVAL rather than letting the VM
+    // layer produce a harder-to-diagnose error.
+    if flags.contains(MremapFlags::MREMAP_FIXED) {
+        if !new_addr.is_multiple_of(PAGE_SIZE) {
+            return_errno_with_message!(Errno::EINVAL, "mremap: `new_addr` must be page-aligned");
+        }
+        if !is_userspace_vaddr_range(new_addr, new_size) {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "mremap: `new_addr` is not in the userspace range"
```
Comment (cchanging) [inline kernel/src/syscall/mremap.rs]: We can use `__log_prefix` to add this prefix consistently.

File: kernel/src/syscall/mremap.rs [hunk 6]
```diff
@@ -78,10 +123,8 @@ fn do_sys_mremap(
         // the old mapping, it is necessary to check whether the old range lies
         // in a single mapping.
         //
-        // FIXME: According to <https://man7.org/linux/man-pages/man2/mremap.2.html>,
-        // if the `MREMAP_MAYMOVE` flag is not set, and the mapping cannot
-        // be expanded at the current `Vaddr`, we should return an `ENOMEM`.
```
Comment (cchanging) [inline kernel/src/syscall/mremap.rs:83]: We can simply remove this `FIXME`.

File: kernel/src/vm/vmar/vmar_impls/mod.rs [hunk 7]
```diff
@@ -7,6 +7,7 @@ pub(super) mod page_fault;
 mod protect;
 mod query;
 mod remap;
+pub use remap::RemapOldMappingAction;
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/mod.rs]: Don't put this export statement among the `mod` declarations.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 8]
```diff
@@ -59,16 +68,25 @@ impl Vmar {
     /// - If `new_addr` is `None`, a new range of size `new_size` will be
     ///   allocated, and the original mapping will be moved there.
     ///
+    /// When `action` is [`RemapOldMappingAction::Keep`], the old mapping is
+    /// kept as an anonymous zero-fill-on-demand mapping after the pages have
+    /// been moved to the new address.  In this case `new_size` must equal
+    /// `old_size`.  If `new_addr` is `None`, the kernel automatically allocates
+    /// a new address range; otherwise, the mapping is moved to `new_addr`
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: This document is outdated.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 9]
```diff
@@ -131,10 +149,13 @@ impl Vmar {
                 &mut rss_delta,
             )?
         } else {
-            debug_assert!(new_size > old_size);
+            debug_assert!(new_size > old_size || action == RemapOldMappingAction::Keep);
 
-            // Fast path: expand the old mapping in place to the new size
-            if is_userspace_vaddr_range(old_addr, new_size)
+            // Fast path: expand the old mapping in place to the new size.
+            // Skip when `action` is `Keep` since we must actually move pages
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: `Keep` -> `RemapOldMappingAction::Keep`.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 10]
```diff
@@ -212,6 +243,14 @@ impl Vmar {
         cursor.flusher().dispatch_tlb_flush();
         cursor.flusher().sync_tlb_flush();
 
+        // For Unmap: old_mapping was removed from the tree earlier and has no
+        // references.  For Keep (MREMAP_DONTUNMAP): the old mapping remains in
+        // the tree; its physical pages are gone, but its VmMapping retains the
+        // original file-backing, permissions, and flags.  Subsequent page faults
+        // at `old_addr` will be handled by the existing VmMapping (anonymous
+        // mappings get zero-filled pages, file-backed mappings read from the
+        // page cache).
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: The position of this comment is weird. I think you can explain the detail semantics in the documents of `RemapOldMappingAction`

File: kernel/src/vm/vmar/mod.rs [hunk 11]
```diff
@@ -25,7 +27,7 @@ pub fn is_userspace_vaddr(vaddr: Vaddr) -> bool {
 }
 
 /// Returns whether `vaddr` and `len` specify a legal user space virtual address range.
-fn is_userspace_vaddr_range(vaddr: Vaddr, len: usize) -> bool {
+pub(crate) fn is_userspace_vaddr_range(vaddr: Vaddr, len: usize) -> bool {
```
Comment (cchanging) [inline kernel/src/vm/vmar/mod.rs]: `pub(crate)` -> `pub`.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 12]
```diff
@@ -149,22 +170,32 @@ impl Vmar {
...
-                inner.insert_without_try_merge(right);
-            }
-            old_mapping
-        };
-        // Note that we have ensured that `new_size >= old_size` at the beginning.
-        let new_mapping = old_mapping.clone_for_remap_at(new_range.start);
-        inner.insert_try_merge(new_mapping.enlarge(new_size - old_size));
+            let old_ref = inner.vm_mappings.find_one(&old_mapping_addr).unwrap();
+            let new_mapping = old_ref.clone_for_remap_at(new_range.start);
+            inner.insert_try_merge(new_mapping.enlarge(new_size - old_size));
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: I think most operations should be closer to the `else` branch—for example, `split_range`? Since I see the CI is currently failing, I suspect there may still be some bugs in the logic here.

File: kernel/src/syscall/mremap.rs (hunk 4 above)
Comment (laipuran) [reply kernel/src/syscall/mremap.rs:64]: Yes, it can be moved to fit the pattern.

File: kernel/src/syscall/mremap.rs (hunk 5 above)
Comment (laipuran) [reply kernel/src/syscall/mremap.rs]: I found all the other files are not using the prefix, so I will remove them to be the same.

File: kernel/src/vm/vmar/vmar_impls/mod.rs [hunk 13]
```diff
@@ -240,15 +240,23 @@ impl VmarInner {
         let mut vm_mapping = vm_mapping;
         let addr = vm_mapping.map_to_addr();
 
-        if let Some(prev) = self.vm_mappings.find_prev(&addr) {
+        if let Some(prev) = self
+            .vm_mappings
+            .find_prev(&addr)
+            .filter(|prev| !util::is_intersected(&vm_mapping.range(), &prev.range()))
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/mod.rs]: Why add this filter? If the found `vm_mapping` is intersected with the `prev`, there should be something wrong during the procedure of building mappings or resize operations. You should fix those bugs instead of adding this filter.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 14]
```diff
@@ -5,6 +5,18 @@ use ostd::{mm::vm_space::VmQueriedItem, task::disable_preempt};
...
 
+/// Controls how the old mapping is handled during a `remap` operation.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RemapOldMappingAction {
+    /// Remove the old mapping after moving pages.
+    Unmap,
+    /// Keep the old address range mapped with its original properties.
+    /// Its physical pages are moved away, so subsequent page faults at
+    /// the old address will allocate fresh pages (zero-filled for anonymous
+    /// mappings, from the page cache for file-backed mappings).
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: "(zero-filled for anonymous mappings, from the page cache for file-backed mappings)" This sentence is a bit wordy. You can simply remove it since you have said "with its original properties".

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 15]
```diff
@@ -59,16 +71,28 @@ impl Vmar {
...
     ///
+    /// When `action` is [`RemapOldMappingAction::Keep`], the old mapping
+    /// stays in the tree with its original properties (file-backing,
+    /// permissions, flags).  Its physical pages are moved away, so
+    /// subsequent page faults at the old address will allocate fresh pages
+    /// (zero-filled for anonymous mappings, from the page cache for
+    /// file-backed mappings).  In this case `new_size` must equal `old_size`.
+    /// If `new_addr` is `None`, the kernel automatically allocates a new
+    /// address range; otherwise, the mapping is moved to `new_addr` as with
+    /// `MREMAP_FIXED`.
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: This document contains the details which is duplicate with the document of `RemapOldMappingAction::Keep`. Simplify it.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 16]
```diff
@@ -131,16 +155,24 @@ impl Vmar {
                 &mut rss_delta,
             )?
         } else {
-            debug_assert!(new_size > old_size);
+            debug_assert!(
+                new_size > old_size && action == RemapOldMappingAction::Unmap
+                    || new_size == old_size
+            );
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: This `debug_assert` is unnecessary. We have guaranteed `old_size <= new_size`. 

If you want to add a `debug_assert`, I think you only need to add a `debug_assert` for `action == RemapOldMappingAction::Unmap` in the two branches where `new_size > old_size` and `new_size < old_size` and add some explaination of requirements in the document.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 17]
```diff
@@ -149,22 +181,57 @@ impl Vmar {
...
-            let (left, old_mapping, right) = vm_mapping.split_range(&old_range);
+            let vm_mapping = inner
+                .remove(&old_mapping_addr)
+                .expect("mapping must exist after check_lies_in_single_mapping");
+            if !vm_mapping.is_private_anonymous() {
+                return_errno_with_message!(
+                    Errno::EINVAL,
+                    "MREMAP_DONTUNMAP requires a private anonymous mapping"
+                );
+            }
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: From the manual:

> Before Linux 5.13, the MREMAP_DONTUNMAP flag could be used only with private anonymous mappings

Didn’t you run the file-backed tests before? So does that mean this check isn’t needed on Linux 6.*?

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 18]
```diff
@@ -149,22 +181,57 @@ impl Vmar {
...
+                if let Some(left) = left {
+                    inner.insert_without_try_merge(left);
+                }
+                if let Some(right) = right {
+                    inner.insert_without_try_merge(right);
+                }
+                old_mapping
+            };
+            let new_mapping = old_mapping.clone_for_remap_at(new_range.start);
+            inner.insert_try_merge(new_mapping.enlarge(new_size - old_size));
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: Many duplicate logics with the branch before.

File: kernel/src/syscall/mremap.rs (hunk 6 above)
Comment (cchanging) [reply kernel/src/syscall/mremap.rs:83]: I mean just remove 

> // When the mapping cannot be expanded at the current Vaddr,
        // `resize_mapping` correctly returns `ENOMEM` as required by the man page.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 19]
```diff
@@ -131,16 +155,24 @@ impl Vmar {
...
+                && is_userspace_vaddr_range(old_addr, new_size)
                 && inner
                     .alloc_free_region_exact(old_range.end, new_size - old_size)
                     .is_ok()
             {
                 let old_mapping_addr = inner.check_lies_in_single_mapping(old_addr, old_size)?;
-                let old_mapping = inner.remove(&old_mapping_addr).unwrap();
+                let old_mapping = inner
+                    .remove(&old_mapping_addr)
+                    .expect("mapping must exist after check_lies_in_single_mapping");
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: -> `check_lies_in_single_mapping`. In fact in cases like this, where the guarantee is obvious and made very close by, it’s totally fine to just `unwrap()` directly.

File: kernel/src/syscall/mremap.rs [hunk 20]
```diff
@@ -28,33 +31,77 @@ fn do_sys_mremap(
...
+    // `new_addr`.  Reject early with EINVAL rather than letting the VM
+    // layer produce a harder-to-diagnose error.
+    if flags.contains(MremapFlags::MREMAP_FIXED) {
+        if !new_addr.is_multiple_of(PAGE_SIZE) {
+            return_errno_with_message!(Errno::EINVAL, "`new_addr` must be page-aligned");
+        }
+        if !is_userspace_vaddr_range(new_addr, new_size) {
+            return_errno_with_message!(Errno::EINVAL, "`new_addr` is not in the userspace range");
+        }
+    }
```
Comment (cchanging) [inline kernel/src/syscall/mremap.rs]: I’m not sure this fast path is meaningful, because there isn’t really any time-consuming or side-effecting operation between this point and the same check in `Vmar::remap`. Would this affect the ordering of which `errno` gets returned?

File: kernel/src/syscall/mremap.rs [hunk 21]
```diff
@@ -28,33 +31,77 @@ fn do_sys_mremap(
...
     if old_size == 0 {
-        return_errno_with_message!(
-            Errno::EINVAL,
-            "mremap: copying shareable mapping is not supported"
-        );
+        return_errno_with_message!(Errno::EINVAL, "copying shareable mapping is not supported");
+    }
+    if !is_userspace_vaddr(old_addr) {
+        return_errno_with_message!(Errno::EFAULT, "`old_addr` is not in the userspace range");
     }
```
Comment (cchanging) [inline kernel/src/syscall/mremap.rs]: Same question.

Comment (laipuran) [reply kernel/src/syscall/mremap.rs]: `is_userspace_vaddr(old_addr)` returns an explicit EFAULT. Without it, an invalid `old_addr` would reach `Vmar::remap`, where find_one returns None for any unmapped address — it cannot tell whether the address is genuinely outside userspace or simply unmapped. The syscall boundary is where this distinction belongs.

**Comment by laipuran** [comment]
Thanks for giving patient instructions!

Made changed as you demanded, and refined documents.

Except the check for `old_addr`, the reason is commented above.

Have a nice day!

File: book/src/kernel/linux-compatibility/syscall-flag-coverage/memory-management/mremap.scml
Comment (cchanging) [inline book/src/kernel/linux-compatibility/syscall-flag-coverage/memory-management/mremap.scml:16]:
```suggestion
// Resize an existing memory mapping. Relocation is allowed if given `MREMAP_MAYMOVE`.
mremap(
old_address,
old_size,
new_size,
flags = MREMAP_DONTUNMAP | MREMAP_MAYMOVE
);

// Resize an existing memory mapping and force relocation to a specified location.
mremap(
old_address,
old_size,
new_size,
flags = MREMAP_DONTUNMAP | MREMAP_MAYMOVE | MREMAP_FIXED,
new_address
);
```

File: kernel/src/syscall/mremap.rs [hunk 22]
```diff
@@ -28,33 +31,65 @@ fn do_sys_mremap(
...
-        "mremap: old_addr = 0x{:x}, old_size = {}, new_size = {}, flags = {:?}, new_addr = 0x{:x}",
+        "old_addr = 0x{:x}, old_size = {}, new_size = {}, flags = {:?}, new_addr = 0x{:x}",
         old_addr, old_size, new_size, flags, new_addr,
     );
 
+    let action = if flags.contains(MremapFlags::MREMAP_DONTUNMAP) {
+        RemapOldMappingAction::Keep
+    } else {
+        RemapOldMappingAction::Unmap
+    };
```
Comment (cchanging) [inline kernel/src/syscall/mremap.rs]: Move this declaration close to its first usage.

File: kernel/src/syscall/mremap.rs (hunk 21 above)
Comment (cchanging) [reply kernel/src/syscall/mremap.rs]: old_addr -> `old_addr`.

> where find_one returns None for any unmapped address — it cannot tell whether the address is genuinely outside userspace or simply unmapped.

Yes, but if `find_one` fails it will also return `EFAULT`. I just worry about the ordering of the returned error number. Have you tested the problem? If not, can you test calling `mremap` with an `old_addr` which is out of userspace, and the `flag` contains `MREMAP_FIXED` but does not contain `MREMAP_MAYMOVE`? Before this modification, `mremap` will return `EINVAL` in this case, and now it will return `EFAULT`. You can try this test in Linux to find the answer.

File: kernel/src/vm/vmar/vmar_impls/remap.rs [hunk 23]
```diff
@@ -149,21 +167,26 @@ impl Vmar {
...
+
+        if action == RemapOldMappingAction::Keep {
+            // Keep the middle mapped at its original address (pages become
+            // zero-fill-on-demand after the move below).
+            inner.insert_without_try_merge(middle.clone_for_remap_at(old_range.start));
+        }
+
+        // Place a clone at the target address; pages are moved below.
+        let new_mapping = middle.clone_for_remap_at(new_range.start);
         inner.insert_try_merge(new_mapping.enlarge(new_size - old_size));
```
Comment (cchanging) [inline kernel/src/vm/vmar/vmar_impls/remap.rs]: I just found there shouldn’t be any need to split here in `Keep` mode. The original mapping in `Keep` mode should not end up being stored in a split form? You can confirm this first.

If the original mapping does not need to be splited, I think we may need an additional API like `clone_range_for_remap_at`:

```rust
let old_mapping_addr = inner.check_lies_in_single_mapping(old_addr, old_size)?;

// Create a new `VmMapping`.
let new_mapping = if action == RemapOldMappingAction::Unmap {
    let vm_mapping = inner.remove(&old_mapping_addr).unwrap();
    let (left, old_mapping, right) = vm_mapping.split_range(&old_range);
    if let Some(left) = left {
        inner.insert_without_try_merge(left);
    }
    if let Some(right) = right {
        inner.insert_without_try_merge(right);
    }
    // Note that we have ensured that `new_size >= old_size` at the beginning.
    old_mapping.clone_for_remap_at(new_range.start)
} else {
    let old_mapping = inner.vm_mappings.find_one(&old_mapping_addr).unwrap();
    old_mapping.clone_range_for_remap_at(old_addr..old_addr+old_size, new_range.start)
};

inner.insert_try_merge(new_mapping.enlarge(new_size - old_size));
```

Take this as a reference.

File: kernel/src/vm/vmar/mod.rs [hunk 24]
```diff
@@ -25,7 +27,7 @@ pub fn is_userspace_vaddr(vaddr: Vaddr) -> bool {
 }
 
 /// Returns whether `vaddr` and `len` specify a legal user space virtual address range.
-fn is_userspace_vaddr_range(vaddr: Vaddr, len: usize) -> bool {
+pub fn is_userspace_vaddr_range(vaddr: Vaddr, len: usize) -> bool {
```
Comment (cchanging) [inline kernel/src/vm/vmar/mod.rs]: Now this modification is not necessary, just keep `fn`.

File: test/initramfs/src/regression/memory/mmap/mmap_and_mremap.c [hunk 25]
```diff
@@ -205,3 +205,159 @@ FN_TEST(mremap_may_fail_after_unmap)
...
+	TEST_RES(strcmp(new_addr, "last"), _ret == 0);
+	TEST_RES(addr[2 * PAGE_SIZE], _ret == '\0');
+
+	// The first page is untouched.
+	TEST_RES(strcmp(addr, "unchanged"), _ret == 0);
+
+	TEST_SUCC(munmap(new_addr, PAGE_SIZE));
+	TEST_SUCC(munmap(addr, 3 * PAGE_SIZE));
+}
+END_TEST()
```
Comment (cchanging) [inline test/initramfs/src/regression/memory/mmap/mmap_and_mremap.c]: We should not mention the kernel code details like (`split_range` returns `left`=Some, `middle`, `right`=None). in the C test comments. Besides, if `MREMAP_DONTUNMAP` does not need to do split operation, these two tests can be removed and we should also adjust `mremap_dontunmap_trim`.

File: kernel/src/vm/vmar/vmar_impls/remap.rs (hunk 23 above)
Comment (laipuran) [reply kernel/src/vm/vmar/vmar_impls/remap.rs]: OK, thanks

File: kernel/src/syscall/mremap.rs (hunk 21 above)
Comment (laipuran) [reply kernel/src/syscall/mremap.rs]: OK, deleting the check is the best choice

File: book/src/kernel/linux-compatibility/syscall-flag-coverage/memory-management/mremap.scml [hunk 26]
```diff
@@ -6,11 +6,21 @@ mremap(
...
-// Resize an existing memory mapping and force relocation to a specified location.
+// Relocate an existing memory mapping while keeping the old mapping
+// as anonymous zero-fill-on-demand.
 mremap(
     old_address,
     old_size,
     new_size,
-    flags = MREMAP_MAYMOVE | MREMAP_FIXED,
+    flags = MREMAP_DONTUNMAP | MREMAP_MAYMOVE
+);
```
Comment (cchanging) [inline book/src/kernel/linux-compatibility/syscall-flag-coverage/memory-management/mremap.scml:7]: Merge these blocks.

File: kernel/src/syscall/mremap.rs [hunk 27]
```diff
@@ -63,25 +92,20 @@ fn do_sys_mremap(
...
         } else {
-            vmar.remap(old_addr, old_size, None, new_size)
+            vmar.remap(old_addr, old_size, None, new_size, action)
         }
     } else {
         if flags.contains(MremapFlags::MREMAP_FIXED) {
             return_errno_with_message!(
                 Errno::EINVAL,
-                "mremap: `MREMAP_FIXED` specified without also specifying `MREMAP_MAYMOVE`"
+                "MREMAP_FIXED specified without also specifying MREMAP_MAYMOVE"
```
Comment (cchanging) [inline kernel/src/syscall/mremap.rs:103]: -> `MREMAP_FIXED`
-> `MREMAP_MAYMOVE`

File: test/initramfs/src/regression/memory/mmap/mmap_and_mremap.c (hunk 25 above)
Comment (cchanging) [reply test/initramfs/src/regression/memory/mmap/mmap_and_mremap.c]: This has not been resolved.

File: kernel/src/vm/vmar/vm_mapping.rs [hunk 28]
```diff
@@ -130,6 +130,13 @@ impl VmMapping {
         vm_mapping
     }
 
+    pub(super) fn clone_range_for_remap_at(&self, range: Range<Vaddr>, va: Vaddr) -> VmMapping {
+        let mut vm_mapping = self.new_fork();
+        vm_mapping.map_to_addr = va;
+        vm_mapping.map_size = NonZeroUsize::new(range.end - range.start).unwrap();
+        vm_mapping
```
Comment (cchanging) [inline kernel/src/vm/vmar/vm_mapping.rs]: We can not directly call `new_fork` and `dup` the `mapped_mem` here, because the new mapping may have different offset to the backend compare to the original mapping.

```rust
pub(super) fn clone_range_for_remap_at(&self, range: Range<Vaddr>, va: Vaddr) -> VmMapping {
        let new_mapped_mem = self.mapped_mem.clone_range(range);
        VmMapping {
            map_size: NonZeroUsize::new(range.end - range.start).unwrap(),
            map_to_addr: va,
            mapped_mem: self.mapped_mem.clone_range(range),
            path: self.path.clone(),
            ..*self
        }
}

impl MappedMemory {
    /// Duplicates the mapped memory capability for a sub-range.
    ///
    /// The `range` is relative to the start of the mapping that owns this
    /// [`MappedMemory`].
    fn clone_range(&self, range: Range<Vaddr>) -> Self {
        debug_assert!(!range.is_empty());
        debug_assert!(range.start.is_multiple_of(PAGE_SIZE));
        debug_assert!(range.end.is_multiple_of(PAGE_SIZE));

        match self {
            MappedMemory::Anonymous => MappedMemory::Anonymous,
            MappedMemory::Vmo(vmo) => {
                let offset = vmo
                    .offset()
                    .checked_add(range.start)
                    .expect("mapped VMO offset should not overflow");
                MappedMemory::Vmo(vmo.dup_at_offset(offset))
            }
            MappedMemory::Device => MappedMemory::Device,
        }
    }
}
```

and with `MappedMemory::clone_range`, I think you can also simplify the logics in `VmMapping::split` (though it is out of the scope..)

```rust
    pub fn split(self, at: Vaddr) -> (Self, Self) {
        debug_assert!(self.map_to_addr < at && at < self.map_end());
        debug_assert!(at.is_multiple_of(PAGE_SIZE));

        let left_size = at - self.map_to_addr;
        let right_size = self.map_size.get() - left_size;
        let r_mapped_mem = self.mapped_mem.clone_range(left_size..self.map_size.get());
        let l_mapped_mem = self.mapped_mem;
        let left = Self {
            map_size: NonZeroUsize::new(left_size).unwrap(),
            map_to_addr: self.map_to_addr,
            mapped_mem: l_mapped_mem,
            path: self.path.clone(),
            ..self
        };
        ...
     }
```

File: kernel/src/syscall/mremap.rs (hunk 27 above)
Comment (laipuran) [reply kernel/src/syscall/mremap.rs:103]: I analysed all other files with `return_errno_with_message!`, finding most of them are not using back ticks, so I removed them.

File: book/src/kernel/linux-compatibility/syscall-flag-coverage/memory-management/mremap.scml (hunk 26 above)
Comment (laipuran) [reply book/src/kernel/linux-compatibility/syscall-flag-coverage/memory-management/mremap.scml:7]: Sorry for making this mistake, for I misunderstood the scml pattern.

File: kernel/src/vm/vmar/vm_mapping.rs (hunk 28 above)
Comment (laipuran) [reply kernel/src/vm/vmar/vm_mapping.rs]: I modified and applied your code, it seems to shorten the file and runs well.

File: kernel/src/vm/vmar/vm_mapping.rs [hunk 29]
```diff
@@ -834,6 +837,28 @@ impl MappedMemory {
...
+                let new_offset = vmo
+                    .offset()
+                    .checked_add(offset)
+                    .expect("mapped VMO offset should not overflow");
+                MappedMemory::Vmo(vmo.dup_at_offset(new_offset))
+            }
+            MappedMemory::Device => MappedMemory::Device,
+        }
+    }
 }
```
Comment (cchanging) [inline kernel/src/vm/vmar/vm_mapping.rs:856]: Sorry, I didn’t think through some of the details carefully in the reference I provided yesterday. It would be better to rename this API to `dup_at_offset` for the sake of consistency, and in fact the `size` parameter is not really necessary either.

File: kernel/src/vm/vmar/vm_mapping.rs [hunk 30]
```diff
@@ -834,6 +837,28 @@ impl MappedMemory {
             MappedMemory::Device => MappedMemory::Device,
         }
     }
+
+    /// Duplicates the mapped memory capability for a sub-range.
+    ///
+    /// The `offset` is relative to the start of the mapping, and must be
+    /// page-aligned.
+    pub(super) fn clone_range(&self, offset: usize, size: usize) -> Self {
```
Comment (cchanging) [inline kernel/src/vm/vmar/vm_mapping.rs]: `fn` is enough?

**Review by cchanging — APPROVED** [review]
Some minor comments, otherwise LGTM!

**Comment by laipuran** [comment]
Wait, the function name remains unchanged, let me do a patch.

**Review by taosue — APPROVED** [review]
LGTM! Thanks for your contribution!

**Comment by laipuran** [comment]
OK, finished.

Thanks for reviewing the codes.

Have a nice day!

File: kernel/src/vm/vmar/vm_mapping.rs [hunk 31]
```diff
@@ -653,38 +664,28 @@ impl VmMapping {
...
             ..self
         };
         let right = Self {
             map_size: NonZeroUsize::new(right_size).unwrap(),
             map_to_addr: at,
-            mapped_mem: r_mapped_mem,
+            mapped_mem: r_mapped,
+            path: self.path.clone(),
             ..self
         };
```
Comment (cchanging) [inline kernel/src/vm/vmar/vm_mapping.rs:685]: Oh, one more thing: looking at the reference I provided yesterday, `l_mapped` here doesn’t need `dup_at_offset`—it can just use the original `self.mapped_mem` directly, which avoids an extra `clone`. The same goes for the `clone` of `path` on line 686; it doesn’t seem necessary. Also, I think the original names `l_mapped_mem` and `r_mapped_mem` were quite good, so maybe there’s no need to change them.

Comment (laipuran) [reply kernel/src/vm/vmar/vm_mapping.rs:685]: OK, I will fix that.

File: kernel/src/vm/vmar/vm_mapping.rs [hunk 32]
```diff
@@ -684,7 +683,6 @@ impl VmMapping {
             mapped_mem: r_mapped_mem,
             ..self
         };
-
```
Comment (cchanging) [inline kernel/src/vm/vmar/vm_mapping.rs:687]: This blank line should not be removed

Comment (laipuran) [reply kernel/src/vm/vmar/vm_mapping.rs:687]: Fixed that.

**Review by StevenJiang1110 — APPROVED** [review]
LGTM! Thanks for your contribution!

**Review by cqs21 — APPROVED** [review]
LGTM! Thanks for your contribution!
